### PR TITLE
Ensembling support in generation for models with different output softmax size

### PR DESCRIPTION
**Pull request by liezl200**
Summary:
In general, this diff enables the ensembling of multiple models with different sizes of output softmax (for example where model 1 outputs a set of 10 translation candidates + 10 probabilities for each of those candidates, and model 2 outputs a set of 20 translation candidates + 20 probabilities for each of those candidates). See documentation in comments for a full example.

The context of this diff is to allow ensembling of multiple models trained with word prediction vocab reduction.

Differential Revision: D9504725

### CODE REVIEW

File: pytorch_translate/beam_decode.py [hunk 1]
```diff
@@ -517,9 +530,95 @@ def _encode(self, encoder_input, reorder_indices):
...
+        possible_translation_tokens: [0, 3, 5]
+        probs: [0.4, 0.5, 0.1]
+
+        all_translation_tokens: [[3, 7, 8, 9], [0, 3, 5]]
+        all_probs: [[0.25, 0.25, 0.25, 0.25], [0.4, 0.5, 0.1]]
+        possible_translation_tokens = [0, 3, 5, 7, 8, 9] (order varies)
+        mapped_probs for model 1: [0  , 0.25, 0  , 0.25, 0.25, 0.25]
+        mapped_probs for model 2: [0.4, 0.5 , 0.1, 0   , 0   , 0]
+
+        avg_probs = [0.4, 0.75, 0.1, 0.25, 0.25, 0.25] (order varies but
```
Comment (jmp84) [inline pytorch_translate/beam_decode.py]: This is the sum, not the average.

Comment (jmp84) [reply pytorch_translate/beam_decode.py]: Could you clarify in doc whether this is summing log prob?

Comment (jhcross) [reply pytorch_translate/beam_decode.py]: In practice the values are log probs, and they are just summed, not averaged. This should certainly be clearer in the code (misleading variable names don't help), but that's mostly unrelated to this change.
